### ISSUE TITLE
[API Proposal]: Provide path information for custom JsonException error messages in System.Text.Json

### ISSUE THREAD

**Issue by magyarandras** [dotnet/runtime#79755]
### Background and motivation

Currently, the path information (Path, LineNumber, BytePositionInLine) is only appended to the error message when no custom error message is provided in the constructor of JsonException. However, when a custom message is set, there is no "straightforward way" to add it optionally. If someone throws a JsonException from the Read method of a custom JsonConverter, this information isn't available on the Utf8JsonReader.

As a workaround, it is possible to catch the exception at a "higher level" and add that information, but I think there should be an easier way to handle this. Furthermore, in this case, the author of a custom JsonConverter has to rely on someone to catch the exception and append or insert the path information later.

A public property could be used to specify whether the path information should be inserted into the custom error message using a format string. This way the path information would be both optional (sometimes it may not be wanted in the custom error message) and flexible.

### API Proposal

```csharp
public class JsonException : Exception
{
    /// <summary>
    /// Specifies that 'try' logic should insert Path information to the exception message using a format string.
    /// </summary>
    public bool InsertPathInformation { get; init; }
}
```
For more information see my implementation: [https://github.com/magyarandras/runtime/commit/f56f9c4a5cf807e4bacdcc1ea2e0327473457d49](https://github.com/magyarandras/runtime/commit/f56f9c4a5cf807e4bacdcc1ea2e0327473457d49)

### API Usage

```csharp
public class CustomJsonConverterWithInsertPathInfo : JsonConverter<int>
{
    public override int Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => throw new JsonException("Custom error message Path: {0}, Line number: {1}, Byte position in line: {2}") { InsertPathInformation = true };

    public override void Write(Utf8JsonWriter writer, int value, JsonSerializerOptions options) => throw new NotImplementedException();
}

public class PropertyUsingCustomConverterThrowingWithPathInfo
{
    [JsonConverter(typeof(CustomJsonConverterWithInsertPathInfo))]
    public int Id { get; set; }
}
```

```csharp
try
{
    JsonSerializer.Deserialize<PropertyUsingCustomConverterThrowingWithPathInfo>(@"{ ""Id"": ""invalid_value"" }");
}
catch (JsonException ex)
{
    //Should be "Custom error message Path: $.Id, Line number: 0, Byte position in line: 23"
    Console.WriteLine(ex.Message);
}
```

### Alternative Designs

- A constructor parameter can be used instead of the init-only setter.
- Another possibility is to provide the information to the format string in every case without the bool flag, in that case, there is no need for public API change, however, it could break people's existing code if they use format strings in the error messages.

**Comment by layomia** [comment]
Yes we could provide an opt-in mechanism to include the path (i.e `JsonException.Path`) produced by the serializer or an API to for users to set it.

@steveharter mentioned that the thought during API review was that the user may have a specific error message already and might not want the exception info cluttered.

Depending on user feedback here we might want to challenge that assumption and perhaps include path info by default.

**Comment by singggum3b** [comment]
Might not be related but `System.Text.Json` have a `String.resx` file, but it seem impossible to override this file from application because of the hardcoded `SR` class. Any idea how ? 
Sorry i'm new to Dotnet.
